### PR TITLE
[Model] Allow loading from original Mistral format

### PR DESCRIPTION
**Pull request by patrickvonplaten**
Mistral models are usually uploaded in two formats:
- The original consolidated format:
[consolidated.safetensors](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3/blob/main/consolidated.safetensors)
[params.json](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3/blob/main/params.json)

- A bit later the HF format is usually also uploaded:
[model-00001-of-00003.safetensors](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3/blob/main/model-00001-of-00003.safetensors)
[config.json](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3/blob/main/config.json)

This PR allows to load directly from the original consolidated format which should make it easier to directly get new models working in VLLM.

The PR can be tested with:

```
vllm serve mistralai/Mistral-Nemo-Instruct-2407 --tokenizer_mode mistral --max_model_len 8192 --load_format consolidated
```

or:

```py
from vllm import LLM, SamplingParams

prompts = [
    "Hello, my name is",
    "The president of the United States is",
    "The capital of France is",
    "The future of AI is",
]
sampling_params = SamplingParams(temperature=0.8, top_p=0.95, seed=0)

model_path = "mistralai/Mistral-7B-Instruct-v0.3"
for model in [model_path]:  # or "mistralai/Mistral-7B-Instruct-v0.3"
    llm = LLM(model=model, tokenizer_mode="mistral", max_model_len=8192, load_format="consolidated")
    outputs = llm.generate(prompts, sampling_params)

    for output in outputs:
        prompt = output.prompt
        generated_text = output.outputs[0].text
        print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}"
```

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/model_loader/weight_utils.py
Comment (mgoin) [inline vllm/model_executor/model_loader/weight_utils.py:254]: super nit: I think it makes more sense to have cache_dir after index_file, similar to how hf_hub_download is called

File: vllm/transformers_utils/config.py
Comment (mgoin) [inline vllm/transformers_utils/config.py:113]: Is there a reason why you have this new config? It seems to have the same information you would have in the config.json, just named differently

File: vllm/model_executor/models/llama.py
Comment (mgoin) [inline vllm/model_executor/models/llama.py:396]: Is this a standard naming scheme used by Llama models as well? I have only seen Mistral models with these style of checkpoints

**Review by mgoin — COMMENTED** [review]
I think this is reasonable overall but are these consolidated checkpoints going to be actively used going forward? 
We tend to want to conform to HF as a standard, so if these checkpoints don't load in transformers now then I think the community will continue to have interest in having those be converted to canonical "HF-style" checkpoints

File: vllm/model_executor/models/llama.py
Comment (patrickvonplaten) [reply vllm/model_executor/models/llama.py:396]: Yeah the original LLama checkpoints have this naming as well: https://github.com/meta-llama/llama/blob/8fac8befd776bc03242fe7bc2236cdb41b6c609c/llama/model.py#L207

(guess most people use the HF format indeed though)

File: vllm/transformers_utils/config.py
Comment (patrickvonplaten) [reply vllm/transformers_utils/config.py:113]: The main reason is because the original format is always stored in [params.json](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3/blob/main/params.json) which accompanies the consolidated.safetensors checkpoints.

Guess there are two problems with config.json

- 1. They are tied quite heavily to transformers. So if one wants to add a new, no transformers-formatted model it doesn't make too much sense to follow transformers config style
- 2. They tend to be quite bloated, e.g. has info that's outdated such as `sliding_window` here: https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3/blob/e0bc86c23ce5aae1db576c8cca6f06f1f73af2db/config.json#L19

File: vllm/model_executor/model_loader/weight_utils.py
Comment (patrickvonplaten) [reply vllm/model_executor/model_loader/weight_utils.py:254]: happy to change

**Comment by patrickvonplaten** [comment]
> I think this is reasonable overall but are these consolidated checkpoints going to be actively used going forward? We tend to want to conform to HF as a standard, so if these checkpoints don't load in transformers now then I think the community will continue to have interest in having those be converted to canonical "HF-style" checkpoints

Hey @mgoin,

Yes at Mistral we're usually actually only releasing the consolidated checkpoints to begin with (see table [here](https://github.com/mistralai/mistral-inference?tab=readme-ov-file#model-download)). The community then often converts the models to transformers but this takes usually some time and then it takes even more time to wait for the next transformers release.

I understand that it's easier to just conform to HF-style, but I guess VLLM also has GGUF support and it could make sense to start allowing more and more support for other formats?

The biggest reason why I think it makes sense to allow also for the consolidated format is so that users could have day-0 inference support for Mistral models for future model releases. VLLM is essentially implementing all the model architectures natively in the model register, so it's just about "correctly" loading the file format no? Guess in the future VLLM can run any nn.Module regardless of whether it's in transformers or not no? It's a bit painful to have to adapt mistral format to transformers format to allow inference in VLLM

File: vllm/transformers_utils/config.py
Comment (timlacroix) [inline vllm/transformers_utils/config.py:149]: this should only be set if not in config dict ? or am I missing sthing ?

File: vllm/transformers_utils/config.py
Comment (timlacroix) [inline vllm/transformers_utils/config.py:152]: same

File: vllm/config.py
Comment (timlacroix) [inline vllm/config.py:750]: CONSOLIDATED is a bit broad as a name no ? Especially since there is SHARDED_STATE above, I think it could lead to confusion.

Thoughts on LoadFormat.MISTRAL ? This makes it clear that the intent is to have us support / maintain the integration of our models into vLLM.

Comment (patrickvonplaten) [reply vllm/config.py:750]: Yes makes sense!

File: vllm/transformers_utils/config.py
Comment (patrickvonplaten) [reply vllm/transformers_utils/config.py:149]: Nice catch - this was actually just a leftover from debugging. Not needed anymore!

File: vllm/config.py
Comment (patrickvonplaten) [reply vllm/config.py]: Changed to "mistral" now - keen to hear what @simon-mo @mgoin think :-)

Comment (robertgshaw2-neuralmagic) [reply vllm/config.py]: +1 on calling it "MISTRAL" format

Comment (mgoin) [reply vllm/config.py]: I agree, it's better to be explicit if possible

**Comment by patrickvonplaten** [comment]
Thanks for the feedback @timlacroix @mgoin - think everything should be incorporated.  Let me know if there is anything else that should be changed!

File: tests/models/test_mistral.py
Comment (mgoin) [inline tests/models/test_mistral.py:41]: I think it would be best if you specifically made a test for "mistralai/Mistral-7B-Instruct-v0.3" where you test the model load with both "safetensors" and "mistral" (and HF to just ensure reference) since both checkpoint formats are present on that model card.

File: vllm/config.py
Comment (mgoin) [inline vllm/config.py:123]: I think you made some good points on supporting checkpoint formats other than HF. Given that assumption, I think it would be best to future-proof the interface of config overrides.

I would like to propose changing this parameter to something like `config_format` that takes a value from a `ConfigFormat` enum, defaulting to ConfigFormat.HF_CONFIG or AUTO. Then `get_config()` can dispatch based on this value to the specified config parsing.

**Review by mgoin — COMMENTED** [review]
I appreciate your explanation and justification @patrickvonplaten ! I think this is in a good place, my last gripe is with `load_params_config` since it would be good to future-proof this for other load formats in the future that may require similar changes. Look in the comments for my proposal

File: vllm/config.py
Comment (patrickvonplaten) [reply vllm/config.py:123]: Makes a lot of sense!

File: tests/models/test_mistral.py
Comment (patrickvonplaten) [reply tests/models/test_mistral.py:41]: That makes sense! Thanks - will take care of it in a bit

File: tests/models/test_mistral.py
Comment (patrickvonplaten) [inline tests/models/test_mistral.py:50]: new test that checks that all tokenizer_mode, load_format and config_format in mistral matches HF format

File: vllm/transformers_utils/config.py
Comment (patrickvonplaten) [inline vllm/transformers_utils/config.py:77]: If we're in auto format and 'config.json' exists we load from HF format

File: vllm/transformers_utils/config.py
Comment (mgoin) [inline vllm/transformers_utils/config.py]: Could you make this an elif where you check for the mistral config, and the else is a failure/warning where we default to HF? I don't think we want to assume something is a mistral config simply if there is no config.json - I'm not sure if HF always requires a specifically named config.json

Comment (patrickvonplaten) [reply vllm/transformers_utils/config.py]: Yeah makes sense - HF does for now always require a config.json (see here: https://github.com/huggingface/transformers/blob/21fac7abba2a37fae86106f87fcf9974fd1e3830/src/transformers/utils/__init__.py#L245)

(can also import this variable)

But yeah will also check for mistral config and else throw an error

Comment (patrickvonplaten) [reply vllm/transformers_utils/config.py]: Made the change - let me know what you think :-)

Comment (mgoin) [reply vllm/transformers_utils/config.py]: Perfect!

**Review by mgoin — APPROVED** [review]
Thanks for working on getting the CI green, looks like you're close. It does seem like AUTO is causing trouble here, so I would be okay with removing it and defaulting to HF if you can't get it stable. Otherwise this LGTM so we can land when green

**Comment by patrickvonplaten** [comment]
> Thanks for working on getting the CI green, looks like you're close. It does seem like AUTO is causing trouble here, so I would be okay with removing it and defaulting to HF if you can't get it stable. Otherwise this LGTM so we can land when green

Yeah I think it should be fine now - added one `is_gguf` statement that fixes everything I believe :-)

**Comment by rosmur** [comment]
will this work for pixtral as well?

**Comment by ryoungj** [comment]
Will this work for the Llama original formats which are in the formats of `consolidated.00.pth` and `params.json`?

**Comment by sfjhu** [comment]
Hi all, I just pulled the latest Docker version and am still getting this error when running Pixtral-12B-2409. Was this fix ever released or am I doing something incorrectly?

**Pull request merged**
